File: 101-109_small_problems/easy_1/sum_of_digits.rb
# PEDAC
# Goal: Write a method that takes one argument, a positive integer, and returns the sum of its digits. For a challenge, try writing this without any basic looping constructs (while, until, loop, and each).

# Identify Expected Input or output
# Make requirements explicit
# Identify rules
# Mental models of problem

# Input: One positive integer
# Output: One positive integer
# No basic looping constructs

# Create examples/test cases
# puts sum(23) => 5
# puts sum(496) => 19
# puts sum(123_456_789) => 45

# Data structures
# Input is an Integer
# Output is an Integer

# Algorithm
# Given a positive integer.
#
# Break up integer into an array.
# Set sum = 0;
# For each integer in array,
#  Add Integer version of integer to sum
# Return sum

# Code

def sum(number)
  numbers = number.to_s.chars

  numbers.reduce {|sum, i| sum.to_i + i.to_i }
end

puts sum(23) == 5
puts sum(496) == 19
puts sum(123_456_789) == 45

File: exercises/js210_small_problems/list_processing/sumOfDigits.js
// Sum of Digits
// Write a function that takes one argument, a positive integer, and returns the sum of its digits. Do this using list processing techniques.
//

// Input: A positive integer
// Output: A sum of its Digits
//
// Algorithm
// 1. Turn number into string
// 2. Split string by Digits
// 3. Map string into numbers
// 4. Use the reduce method to add digits together

function sum(number) {
  return String(number).split('')
                       .map((digit) => Number(digit))
                       .reduce((sum, digit) => sum + digit);
}

// Examples:

console.log(sum(23));           // 5
console.log(sum(496));          // 19
console.log(sum(123456789));    // 45
